src/chrome/crdpMultiplexing/crdpMultiplexor.ts

```diff
@@ -0,0 +1,201 @@
+        return channel;
+    }
+
+    public send(channel: CRDPChannel, data: string): void {
+        const message = JSON.parse(data);
+        if (message.id !== undefined) {
+            message.id = encodifyChannel(channel.id, message.id);
+            data = JSON.stringify(message);
+        } else {
+            throwCriticalError(`Channel [${channel.name}] sent a message without an id: ${data}`);
+        }
+        this._wrappedLikeSocket.send(data);
+    }
+
+    public addListenerOfNonMultiplexedEvent(event: string, cb: Function): void {
+        this._wrappedLikeSocket.on(event, cb);
+    }
+
+    public removeListenerOfNonMultiplexedEvent(event: string, cb: Function): void {
+        this._wrappedLikeSocket.removeListener(event, cb);
+    }
+}
+
+export class CRDPChannel implements LikeSocket {
+    private static timeToPreserveMessagesInMillis = 60 * 1000;
+
+    private _messageCallbacks: Function[] = [];
+    private _enabledDomains: { [domain: string]: boolean } = {};
+    private _pendingMessagesForDomain: { [domain: string]: string[] } = {};
+
+    public callMessageCallbacks(messageData: string): void {
+        this._messageCallbacks.forEach(callback => callback(messageData));
+    }
+
+    public callDomainMessageCallbacks(domain: string, messageData: string): void {
+        if (this._enabledDomains[domain]) {
+            this.callMessageCallbacks(messageData);
+        } else if (this._pendingMessagesForDomain !== null) {
+            // We give clients 60 seconds after they connect to the channel to enable domains and receive all messages
+            this.storeMessageForLater(domain, messageData);
+        }
+    }
+
+    private storeMessageForLater(domain: string, messageData: string): void {
+        let messagesForDomain = this._pendingMessagesForDomain[domain];
+        if (messagesForDomain === undefined) {
+            this._pendingMessagesForDomain[domain] = [];
+            messagesForDomain = this._pendingMessagesForDomain[domain];
+        }
+
+        // Usually this is too much logging, but we might use it while debugging
+        // logger.log(`CRDP Multiplexor - Storing message to channel ${this.name} for ${domain} for later: ${messageData}`);
+        messagesForDomain.push(messageData);
+    }
+
+    constructor(public name: string, public id: number, private _multiplexor: CRDPMultiplexor) { }
+
+    public send(messageData: string): void {
+        const message = JSON.parse(messageData);
+        const method = message.method;
+        const isEnableMethod = method && method.indexOf(".enable") >= 0;
+        let domain;
+
+        if (isEnableMethod) {
+            domain = extractDomain(method);
+            this._enabledDomains[domain] = true;
+        }
+
+        this._multiplexor.send(this, messageData);
+
+        if (isEnableMethod) {
+            this.sendUnsentPendingMessages(domain);
+        }
+    }
+
+    private sendUnsentPendingMessages(domain: string): void {
```

Comment:
You are right, we cannot really sync the state between multiple things, but we are just choosing to ignore the bad side effects that problem brings from the time being. We currently need to enable "Runtime" for all the channels, so we do need to support the same Domain in different channels.